File: BOJ/25_최단경로/11657_타임머신.cpp
#include <cstdio>
#include <iostream>
#include <vector>

/* 
    bellman dist 를 int 로 해줬더니 오버플로우때메 출력초과가 되버림.
    -> 단위 꼭 잘 생각합시다.
*/

#define MAX 500
#define INF 987654321
using namespace std;

int N,M;
vector<pair<int, pair<int,int> > > adjmat;
vector<long long> dist(MAX+1,INF);

bool bellman() {
    bool isCycle = false;
    for(int i=1; i<=N; i++) {
        for(int j=0; j<M; j++) {
            int from = adjmat[j].first;
            int to = adjmat[j].second.first;
            int cost = adjmat[j].second.second;

            if(dist[from] == INF) continue;
            if(dist[to] > dist[from] + cost) {
                if(i == N) isCycle = true;
                dist[to] = dist[from] + cost;
            }
        }
    }
    return !isCycle;
}


int main() {
    int x, y, c;
    scanf("%d %d", &N, &M);
    for (int i = 0; i < M; i++){
        scanf("%d %d %d", &x, &y, &c);
        adjmat.push_back(make_pair(x, make_pair(y, c)));
    }
    dist[1] = 0;
    if (bellman()){
        for (int i = 2; i <= N; i++){
            if (dist[i] == INF) printf("-1\n");
            else printf("%d\n", dist[i]);
        }
    }
    else
        printf("-1\n");
}

File: BOJ/25_최단경로/11657_타임머신.py
import sys

read = sys.stdin.readline
INF = int(1e9)

def bellman() : 

    dist[1] = 0
    isCycle = False
    for i in range(1,N+1):
        for j in range(M) :
            f = adj[j][0]
            t = adj[j][1]
            c = adj[j][2]

            if dist[f] == INF : continue
            if dist[t] > dist[f] + c :
                if i == N : 
                    isCycle = True
                dist[t] = dist[f] + c  
    return isCycle

N,M = map(int,read().split())
dist = [INF for _ in range(N+1)]
adj = []
for _ in range(M) :
    x,y,c = map(int,read().split())
    adj.append((x,y,c))

if bellman() :
    print(-1)
else :
    for i in range(2,N+1) :
        if dist[i] >= INF : 
            print(-1)
        else :
            print(dist[i])
